### ISSUE TITLE
Changes port number to 11024 in Azure

### ISSUE THREAD

**Issue by koistya** [pekim/tedious#262]
After connection is established the original connection string (`config`) is modified, `config.options.port` is set to `11024` thus all subsequent connections using this `config` var fail with a timeout.

This only happens when code is executed in Azure Web App. Wasn't able to reproduce in locally a local environment, connecting to the exact same database.

Perhaps it's related to re-route which may work differently when connection is established from Azure datacenter vs from outside network.

**Comment by arthurschreiber** [comment]
We shouldn't modify options objects when passed in, I've seen the same thing happen as well.

**Comment by alexanderpreston** [comment]
Hi Guys, I have a production application in Azure that is down due to the port number being overridden to 11085. After the override all subsequent connections fail with a timeout.

Is there a quick work around?

Was it caused by the recent Azure SQL Server V12 upgrade?

**Comment by koistya** [comment]
I'm using this workaround for now:

```js
let config = () => Object.assign({}, {/* connection string */});
let connection = new sql.Connection(config());
```

**Comment by lee-houghton** [comment]
If you're getting the error all the time, perhaps you could do some `Object.observe()` magic to figure out where in the tedious code is overwriting the passed-in config? (Unless the Tedious guys already know where it's happening, in which case ignore me.)

**Comment by koistya** [comment]
@lee-houghton I have tried your solution, it prints:

```
config.options.port overwritten:
Trace
    at D:\home\site\wwwroot\server.js:51:13
    at runMicrotasksCallback (node.js:337:7)
    at process._tickCallback (node.js:355:11)
```

**Comment by bretcope** [comment]
As far as I can tell, the only place we would reassign the config port is if we receive a `ENVCHANGE:Routing` token from the server. That's probably still the wrong thing to do since all future connection attempts will go there instead of the originally provided details. However, what is provided back from the server should be a valid re-route.

Not sure if this is what's happening here. It looks like the Tedious implementation of the re-route is correct, but it's hard to know for sure. Perhaps @christopheranderson would be able to help track this down to assess whether the port change is due to an ENVCHANGE token.

**Comment by gleeds** [comment]
I'm seeing the exact same thing.  I upgrade my Azure database to v12, and now only the first connection works in my application running on an Azure Website.  When I run the application locally connected to the production Azure database, it works fine.

**Comment by gleeds** [comment]
I modified Tedious to add some extra logging, I can confirm that the source of the port change is in connect.js line 134.  On the initial connection to the database on my Azure Website, the REROUTING reconnect function fires.  Here is the contents of `this.routingData` (with personal data removed, replaced with all CAPS text)
```
{ protocol: 0,
  port: 11125,
  server: '12HEXDIGITS.tr6.southcentralus1-a.worker.database.windows.net' }
```
And my config after the change:
```
{ driver: 'tedious',
  user: 'clientportal',
  password: 'PASSWORD',
  server: 'DBSERVER-server.database.windows.net',
  database: 'DBSERVER-db',
  options: 
   { encrypt: true,
     connectTimeout: 30000,
     database: 'DBSERVER-db',
     port: 11125,
     requestTimeout: 15000,
     tdsVersion: '7_4',
     rowCollectionOnDone: false,
     rowCollectionOnRequestCompletion: false,
     useColumnNames: false,
     appName: 'node-mssql',
     textsize: '2147483647',
     cancelTimeout: 5000,
     packetSize: 4096,
     isolationLevel: 2,
     cryptoCredentialsDetails: {},
     useUTC: true,
     connectionIsolationLevel: 2,
     readOnlyIntent: false,
     enableAnsiNullDefault: true },
  port: 1433,
  stream: false }
```

One thing that stands out is the `this.routingData` has a different server name, however that server name never makes it back into the config object.

**Comment by gleeds** [comment]
I'm now 90% sure the problem is actually that the updated server name is not persisting but the updated port is, but I'm not sure why this doesn't work.

```
this.config.server = this.routingData.server;
this.config.options.port = this.routingData.port;
```

I'm wondering if it's some weird thing where primitives at the root of `this.config` are being treated as pass-by-value where `this.config.options` is being treated as pass-by-reference (effectively)?

**Comment by gleeds** [comment]
I was right about the issue.  The most straightforward way to fix it seems to be to persist the redirection server name in the options object and to use that server name if it is set.  I'll work on a pull request.

**Comment by arthurschreiber** [comment]
Fixed by #309 and #310. Will be released in `1.11.5` and `1.12.3`.
